ludwig/data/dataset/tfrecord.py

```diff
@@ -0,0 +1,218 @@
+"""TF record dataset."""
+import contextlib
+import glob
+import json
+import math
+
+import os
+import tensorflow as tf
+from ludwig.constants import NAME, PROC_COLUMN
+from ludwig.data.batcher.iterable import IterableBatcher
+from ludwig.data.dataset.base import Dataset
+from ludwig.data.dataset.pandas import PandasDataset
+from ludwig.data.dataset.partitioned import PartitionedDataset
+from ludwig.utils.data_utils import DATA_TRAIN_HDF5_FP
+from ludwig.utils.fs_utils import to_url
+from ludwig.utils.misc_utils import get_combined_features, get_proc_features
+
+
+class TFRecordDataset(Dataset):
+    def __init__(self, url, features, training_set_metadata, compression_type="GZIP"):
+        self.url = to_url(url)[7:]
+        self.features = [feature[PROC_COLUMN] for feature in features]
+        self.training_set_metadata = training_set_metadata
+        self.compression_type = compression_type
+
+        with open(os.path.join(self.url, "meta.json")) as in_file:
+            meta = json.load(in_file)
+        self.size = meta["size"]
+        self.reshape_features = {
+            feature[PROC_COLUMN]: list((-1, *training_set_metadata[feature[NAME]]['reshape']))
+            for feature in features
+            if "reshape" in training_set_metadata[feature[NAME]]
+        }
+
+    def get(self, feature_name, sample):
+        t = sample[feature_name]
+        reshape_dim = self.reshape_features.get(feature_name)
+        if reshape_dim is not None:
+            # When we read a 1D array from disk, we need to reshape it back to its
+            # full dimensions.
+            t = tf.reshape(t, reshape_dim)
+        return t
+
+    def __len__(self):
+        return self.size
+
+    @contextlib.contextmanager
+    def initialize_batcher(self,
+                           batch_size=128,
+                           should_shuffle=True,
+                           seed=0,
+                           ignore_last=False,
+                           horovod=None):
+        cur_shard, shard_count = None, None
+        if horovod:
+            cur_shard, shard_count = horovod.rank(), horovod.size()
+        total_samples = self.size
+        local_samples = int(total_samples / shard_count) if shard_count else total_samples
+
+        # Below are routine optimizations for tf.dataset
+        compression_ext = '.gz' if self.compression_type else ''
+        path = os.path.join(self.url, "*.tfrecords{}".format(compression_ext))
+
+        # Now construct the tfrecrd dataset
+        files = tf.data.Dataset.list_files(path)
+
+        # interleave the tfrecord files for parallel reading
+        dataset = files.interleave(
+            lambda x: tf.data.TFRecordDataset(x, compression_type="GZIP"),
+            num_parallel_calls=tf.data.AUTOTUNE)
+        # Fetch one element so to get the parser.
+        features, feature_lists = self._detect_schema(dataset)
```

Comment:
Instead of detecting the schema here, what if we just stored it in the metadata at write time?